### ISSUE TITLE
Add maxfreespace as a per branch option

### ISSUE THREAD

**Issue by darthShadow** [trapexit/mergerfs#862]
**Is your feature request related to a problem? Please describe.**
Basically, I have 2 remotes merged together, one for smaller files like metadata (posters, artwork etc.) and one for the actual media. If I could specify `maxfreespace` of something like `100M` on the smaller mount, then it could be used for the metadata files and since the media files are larger than 100M, they would go to the bigger mount

**Describe the solution you'd like**
Add `maxfreespace` as a per branch option to facilitate the splitting of metadata & media files.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by trapexit** [comment]
I think I misunderstood you when we spoke prior.

maxfreespace=100M as in the max *file* size? That's not something that is really possible. You don't know the file size before it's written. minfreespace is about how much space is available on a branch... not the files themselves. I'm not sure there is any practical way to do what I think you're looking for.

**Comment by trapexit** [comment]
I suppose a filename / filepath filter of some sort could be used. Some way to say "no files matching this name". Pretty niche and wouldn't be perfect.

**Comment by darthShadow** [comment]
Yeah, I think I explained it incorrectly earlier. Considering there is no way to know the file size (at least that I know of) before it's closed, then this feature won't help in my case.

Filtering would accomplish the required behaviour and that's what I am using right now with rclone (https://rclone.org/filtering/) but you are correct in that it's probably a very niche use-case and probably not worth the time it will take to implement.

Feel free to close the issue if that's the case.

Thanks.

**Comment by trapexit** [comment]
> before it's closed

You can know the size of a file at any time but there isn't a "copy" call. It's an "open", "read/write", "close". And possibly a bunch of other stuff in-between. If you're suggesting checking on close... that's a whole other ball of wax. While the file is being written to it's gotta go somewhere. RAM, a drive, somewhere. To do what you originally suggested would require storing the file temporarily somewhere and then, in effect, rerun the policy to select a location and then move the file. It's possible but it's not so simple because software might be sharing the location or whatnot which would have to carefully considered. Not impossible but awkward in the least.

Regarding filtering. I'm not opposed to adding path filtering to policies. I just want to make sure that would work for you and I will need to figure out the best way to do it. Adding new fields to the branches option needs to be carefully considered. I'm thinking of new ideas for defining it in 3.x but the current syntax isn't really designed for this kind of thing.

**Comment by darthShadow** [comment]
Ok, thanks for the explanation. Temp files are not my ideal solution too.

Yeah, the path filters would work for me but it's not urgent and can be tackled for 3.x as you say since I already have a similar solution with rclone working right now.

**Comment by trapexit** [comment]
What kind of filtering are you looking for? How elaborate? Would fnmatch patterns work? regex?

**Comment by darthShadow** [comment]
I would prefer glob-style matching as used by rclone since it seems like a good compromise between useability & advanced matching without needing to require full regex patterns. 

Docs about rclone filtering: https://rclone.org/filtering/#pattern-syntax

**Comment by trapexit** [comment]
Hmm. I don't recognize that as a standard fnmatch glob. Will have to see what that is and if I can replicate it without needing to write my own matcher.

**Comment by darthShadow** [comment]
You can try out the filtering as a web-hosted version here: https://filterdemo.rclone.org

**Comment by trapexit** [comment]
I'm generally familiar with rclone and the filtering but after reading about it again I don't recognize it as something that is standard. So I'd have to use something else or write my own or find a compatible library. I'd prefer not have to write my own.

**Comment by darthShadow** [comment]
I have asked @ncw to chime in regarding whether it matches any standard matching pattern or whether it's completely custom. If it's custom or we can't find a matching library, then regex will also work for me, it will just increase the complexity of the filters themselves.

**Comment by trapexit** [comment]
I already asked him. I'll look at rsync and git's code. See if they are useful / performant enough for the usecase.

**Comment by nnWhisperer** [comment]
My use case is just mounting a folder to docker, but filtering the files.
Hence, I'm also looking for a way to filter file names, interestingly, it's isn't straightforward unless using rclone/rsync, though I'm not familiar with their performance penalties.
@trapexit Your comment mentioned that you'll benchmark performance. How is the performance of rclone/rsync?

**Comment by trapexit** [comment]
It's not comparable. They aren't analyzing many branches and having to apply that across dozen+ functions.

What exactly is the usecase. Filter what? What behaviors are filtered? There are a dozen+ filesystem calls. Does it affect them all? What exactly does it mean to be filtered? What about creation of temporary files? What about creating new files with the filtered name. There are many things to consider.

**Comment by ncw** [comment]
@trapexit 
> I'm generally familiar with rclone and the filtering but after reading about it again I don't recognize it as something that is standard. So I'd have to use something else or write my own or find a compatible library. I'd prefer not have to write my own.

The rclone filtering is based on file globs roughly like rsyncs but it has some extensions (like the ability to wedge regexps in). So isn't a standard, alas.

It supports `*`, `**`, `?`, `[]`, and `{}`. It has a non-standard extension `{{}}` for regexps. The pattern matches anywhere in the path unless it starts with `/`.

I think if you found a glob library supporting those things (apart from the regexps) and built a filter chain with includes and excludes you would give the user enough power to do anything they wanted.

**Comment by trapexit** [comment]
@ncw I know how it works in rsync and rclone but that is different from a union filesystem. There is (potentitally) a lot more to do than filter readdir output. And even if that was all that was being done having a list of regex/fnmatch like filters... I wouldn't call particularly cheap depending on the specifics. 

I need to know exactly what is to be filtered and how to respond. What if someone tries to create a file with a filtered name? What about rename or link? Unlink? It's not simply "skip file for transfer" like rsync.

**Comment by ncw** [comment]
> I know how it works in rsync and rclone but that is different from a union filesystem. There is (potentitally) a lot more to do than filter readdir output. And even if that was all that was being done having a list of regex/fnmatch like filters... I wouldn't call particularly cheap depending on the specifics.

Rclone has some of these problems too in `rclone mount` which you can apply filters to.

> I need to know exactly what is to be filtered and how to respond. What if someone tries to create a file with a filtered name? What about rename or link? Unlink? It's not simply "skip file for transfer" like rsync.

Rclone only applies filtering to directory listings. If you can create a excluded file, it won't appear in directory listings. That keeps it simple and seems to be good enough.

**Comment by trapexit** [comment]
Right, but this is what I'm trying to get from possible users. Exactly what they are looking for. Because some people want directory listing filtering. Some want policy based filtering (don't consider branch X for files ending in "png"). What happens with a rename when you have a situation where you want it filtering "png" from the path but then a file is renamed to that name?

**Comment by ncw** [comment]
> What happens with a rename when you have a situation where you want it filtering "png" from the path but then a file is renamed to that name?

It will disappear. Not a great result, but there you go!

**Comment by trapexit** [comment]
I meant in a create policy filter. The user wants "don't use this branch for *.png" but that knowledge isn't available till it is already written and the rename happens. You either leave it or move it at rename.
